### PR TITLE
Bump Terraform CLI to v1.5.5

### PR DESCRIPTION
**Pull request by ulucinar**
### Description of your changes

This PR bumps the Terraform CLI version to `v1.5.5` and consumes it from https://github.com/upbound/terraform/releases.

I have:

- [x] Run `make reviewable` to ensure this PR is ready for review.

### How has this code been tested

Manually tested the provider using Terraform `v1.5.5` using the provider example `examples/workspace-inline-aws.yaml`:
```
❯ k get managed
NAME                                    READY   SYNCED   AGE
workspace.tf.upbound.io/sample-inline   True    True     21s
```

Also verified via the the uptest run [here](https://github.com/upbound/provider-terraform/actions/runs/6004142126).

### CODE REVIEW

**Comment by ulucinar** [comment]
/test-examples="examples/workspace-inline-aws.yaml"

File: cluster/images/provider-terraform/Dockerfile [hunk 1]
```diff
@@ -3,14 +3,14 @@ RUN apk --no-cache add ca-certificates bash git curl
...
-ENV TERRAFORM_VERSION=1.5.2
+ENV TERRAFORM_VERSION=1.5.5
 ENV TF_IN_AUTOMATION=1
 ENV TF_PLUGIN_CACHE_DIR=/tf/plugin-cache
 
 ADD "bin/${TARGETOS}_${TARGETARCH}/provider" /usr/local/bin/crossplane-terraform-provider
 ADD .gitconfig .gitconfig
 
-RUN curl -s -L https://releases.hashicorp.com/terraform/${TERRAFORM_VERSION}/terraform_${TERRAFORM_VERSION}_${TARGETOS}_${TARGETARCH}.zip -o terraform.zip \
+RUN curl -s -L https://github.com/upbound/terraform/releases/download/v${TERRAFORM_VERSION}/terraform_${TERRAFORM_VERSION}_${TARGETOS}_${TARGETARCH}.zip -o terraform.zip \
```
Comment (jeanduplessis) [inline cluster/images/provider-terraform/Dockerfile:13]: @bobh66 @ytsarev, we should not change the location of where we pull the Terraform binary from due to the Terraform BSL change. When the time comes where there is a need for a newer version we should evaluate our options before simply upgrading to a new TF version.

Comment (bobh66) [reply cluster/images/provider-terraform/Dockerfile:13]: I would add a comment to that effect before the `RUN` command.

**Review by bobh66 — CHANGES_REQUESTED** [review]
Suggest adding a comment regarding the reason we're pulling from the private repo and why, otherwise someone will change it without realizing.

**Comment by ulucinar** [comment]
/test-examples="examples/workspace-inline-aws.yaml"

**Review by bobh66 — APPROVED** [review]

File: cluster/images/provider-terraform/Dockerfile (hunk 1 above)
Comment (ulucinar) [reply cluster/images/provider-terraform/Dockerfile:16]: Thanks @bobh66, added a comment explaining the situation.

**Pull request merged**